tensorflow-framework/src/main/java/org/tensorflow/framework/losses/Losses.java

```diff
@@ -0,0 +1,707 @@
+      loss = tf.reshape(loss, tf.constant(newShape));
+    }
+    return loss;
+  }
+
+  /**
+   * Computes the squared hinge loss between labels and predictions.
+   *
+   * <p><code>loss = reduceMean(square(maximum(1 - labels * predictions, 0)))</code>
+   *
+   * @param tf the TensorFlow Ops
+   * @param labels true targets, values are expected to be -1 or 1. If binary (0 or 1) labels are *
+   *     provided, they will be converted to -1 or 1.
+   * @param predictions the predictions
+   * @param <T> the data type of the predictions and labels
+   * @return the squared hinge loss
+   */
+  public static <T extends TNumber, U extends TNumber> Operand<T> squaredHinge(
+      Ops tf, Operand<U> labels, Operand<T> predictions) {
+    DataType<T> dataType = predictions.asOutput().dataType();
+    Operand<T> tLabels = cast(tf,  labels, dataType);
+    LossTuple<T> lossTuple = LossesHelper.squeezeOrExpandDimensions(tf, tLabels, predictions, null);
+    predictions = lossTuple.getTarget();
+    tLabels = lossTuple.getLabels();
+    Operand<T> one = cast(tf,  tf.constant(1), dataType);
+    Operand<T> zero = cast(tf,  tf.constant(0), dataType);
+
+    tLabels = maybeConvertLabels(tf, tLabels);
+    return tf.math.mean(
+        tf.math.square(tf.math.maximum(tf.math.sub(one, tf.math.mul(tLabels, predictions)), zero)),
+        tf.constant(-1));
+  }
+
+
+  /**
+   * Smooths binary labels
+   *
+   * @param tf the TensorFlow Ops
+   * @param labels true targets
+   * @param labelSmoothing A number in the range [0, 1]. When 0, no smoothing occurs. When &gt; 0,
+   *     compute the loss between the predicted labels and a smoothed version of the true labels,
+   *     where the smoothing squeezes the labels towards 0.5. Larger values of labelSmoothing
+   *     correspond to heavier smoothing.
+   * @param <T> the data type of the labels
+   * @return the smoothed binary labels
+   */
+  private static <T extends TNumber> Operand<T> smoothLabelsBinaryX(
+      Ops tf, Operand<T> labels, float labelSmoothing) {
+    DataType<T> dataType = labels.asOutput().dataType();
+    Operand<T> oneMinusSmoothing = cast(tf,  tf.constant(1.f - labelSmoothing), dataType);
+    Operand<T> halfSmoothing = cast(tf,  tf.constant(0.5F * labelSmoothing), dataType);
+    return tf.math.add(tf.math.mul(labels, oneMinusSmoothing), halfSmoothing);
+  }
+
+  /**
+   * Smooths categorical labels
+   *
+   * @param tf the TensorFlow Ops
+   * @param labels true targets
+   * @param labelSmoothing Float in <code>[0, 1]</code>. When <code>&gt; 0</code>, label values are smoothed, meaning the
+   *    confidence on label values are relaxed. e.g. <code>label_smoothing=0.2<code> means that we will use a
+   *    value of </code>0.1<code> for label </code>0<code> and </code>0.9<code> for label </code>1<code>
+   * @param <T> the data type of the labels
+   * @return the smoothed categorical labels
+   */
+  private static <T extends TNumber> Operand<T> smoothLabelsCatX(
```

Comment:
Similar comment to above, but `smoothCategoricalLabels`. Also I think the doc should explicitly state that it's smoothing the labels towards `1/n` where `n` is the number of classes.

Reply:
OK